### PR TITLE
[WIP] s3peer: Rewrite to add support for new peer clients 

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
Make it easier to add new s3 peer clients and avoid being limited to only BMS client

## Motivation and Context
Pave the way for new RPC clients for inter nodes communications, like Get Server Info

## How Has This Been Tested?
go test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: cmd/s3-peer-bms-client.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	"fmt"
+	"sync"
+)
+
+// Bucket Meta State interface to be registered in S3 Peers API
+type s3PeerBucketMetaStateAPI struct{}
+
+// Bucket Meta State API name
+func (c s3PeerBucketMetaStateAPI) APIName() string {
+	return "bms"
```
Comment (harshavardhana) [inline cmd/s3-peer-bms-client.go:30]: You should type the API name than using generic string like this, hard to catch bugs.

File: cmd/s3-peer-bms-client.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+// Return Bucket Meta State API for local use
+func (c s3PeerBucketMetaStateAPI) NewLocalClient() interface{} {
```
Comment (harshavardhana) [inline cmd/s3-peer-bms-client.go:34]: Catch all interface should be avoided in the return statement.. The functionaliy which localBucketMetaState would provide should be put into a concrete type.

Comment (harshavardhana) [reply cmd/s3-peer-bms-client.go:34]: A concrete type here should put into an interface like below 

```
type UpdateBucketMetadata interface {
... add relevant methods generic for all metadata updates ...
}
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Overall this is a good change but we need to make sure to concrete types either by writing a common interface compatible type or use a proper struct.

File: cmd/s3-peer-bms-client.go (hunk 2 above)
Comment (vadmeste) [reply cmd/s3-peer-bms-client.go:34]: Actually this PR will help to implement new calls (local and remote) like Get Storage Info (https://github.com/minio/minio/issues/3475).. that's why I am using generic interface{}

Comment (harshavardhana) [reply cmd/s3-peer-bms-client.go:34]: Still we need to explore all possible ways before we can say that we have to use a generic interface{} like this.

**Comment by harshavardhana** [comment]
@vadmeste is this still valid ?

**Comment by vadmeste** [comment]
> @vadmeste is this still valid ?

Not valid anymore since @balamurugana is taking over this

**Pull request closed without merging**